feat: add checks for header case and last character

dzhu opened the pull request:
I came across this project and loved the overall idea and the imperative check, but I noticed that some other common guidelines (restrictions on the leading case and trailing punctuation of the header) were missing; this adds checks to enforce those. The desired case can be specified as either "upper" or "lower"; the punctuation check is implemented as a general check for disallowing the last character in the header to be from any given set of characters.

andrewrynhard commented:
@dzhu Love the changes. Looks like tests are failing though.

andrewrynhard commented on internal/policy/commit/commit.go:
```diff
@@ -33,6 +33,10 @@ type Commit struct {
+	// HeaderCase is the case that the first word of the header must have ("upper" or "lower").
+	HeaderCase string `mapstructure:"headerCase"`
+	// HeaderInvalidLastCharacters is a string containing all invalid last characters for the header.
+	HeaderInvalidLastCharacters string `mapstructure:"headerInvalidLastCharacters"`
```
Should we have a set of default characters?

dzhu replied:
Hmm, how would that show up in config? I played around with the `mapstructure` definitions and I didn't see a good way to expose a default.

andrewrynhard commented on internal/policy/commit/commit.go:
```diff
@@ -33,6 +33,10 @@ type Commit struct {
+	// HeaderCase is the case that the first word of the header must have ("upper" or "lower").
+	HeaderCase string `mapstructure:"headerCase"`
+	// HeaderInvalidLastCharacters is a string containing all invalid last characters for the header.
+	HeaderInvalidLastCharacters string `mapstructure:"headerInvalidLastCharacters"`
```
What do you think of something like:

```
header:
  case: ...
  headerInvalidLastCharacters:
```

to keep all the header related settings together.

andrewrynhard replied:
We could have `body` and `footer` settings later on, and they could follow the same design.

dzhu replied:
That makes a lot of sense; I'll work on it! Should the current `headerLength` and `imperative` also be moved there?

dzhu commented:
I've updated the config structure, including moving the length and imperative checks inside the new subfield. I also went ahead and updated `README.md` and `.conform.yaml` to put in the new checks, since I thought that made sense, but I'm happy to put any of that back if you'd like.

andrewrynhard commented:
@dzhu Thanks! I like all the header options under one struct much better. Since this is a backwards incompatible change, I'll need some time to get things in order before I cut a release that includes this change. Appreciate the time you spent on this!

dzhu commented:
Sounds good, thanks! I had fun working on this.

Roughly speaking, will the time to the release be in the range of days or weeks? I obviously don't want to rush you at all, but I'd like to use the new features at my workplace, and it would be helpful to know for planning. (Looks like we can `go get` a specific commit, so we're not even totally blocked on a release, but the rest of the team might like it a bit better if there is one.)

andrewrynhard commented:
@dzhu looking more like a week. Caught up in another project right now and it’s release week 🙂
